### PR TITLE
Project setup, create and import wallet implementations, create trans…

### PR DESCRIPTION
**Pull request by pabelnl**
…action and query implementations, added dependencies, added podspec

### CODE REVIEW

File: pocket-aion/pocket-aion/PocketAion.swift [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+        
+        // Assign network
+        pocketQuery.network = network
+        pocketQuery.subnetwork = subnetwork
+        
+        return pocketQuery
+    }
+    
+    // MARK: Tools
+    public static func removeJSGlobalObjects() {
```
Comment (luyzdeleon) [inline pocket-aion/pocket-aion/PocketAion.swift]: make private.

File: pocket-aion/pocket-aion/PocketAion.swift [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+        }
+        guard let privateKey = account["privateKey"] as? String else {
+            throw PocketPluginError.walletCreationError("Invalid private key")
+        }
+        
+        guard let address = account["address"] as? String else {
+            throw PocketPluginError.walletCreationError("Invalid address")
+        }
+        
+        return try Wallet(address: address, privateKey: privateKey, network: network, subnetwork: subnetwork, data: nil)
```
Comment (luyzdeleon) [inline pocket-aion/pocket-aion/PocketAion.swift:68]: Unnecesary `try`

File: pocket-aion/pocket-aion/PocketAion.swift [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+        
+        guard let account = jsContext?.objectForKeyedSubscript("account")?.toObject() as? [AnyHashable: Any] else {
+            throw PocketPluginError.transactionCreationError("Failed to create account object")
+        }
+        
+        if account["address"] as? String != publicKey {
+            throw PocketPluginError.transactionCreationError("Failed to create account object")
+        }
+        
+        return try Wallet(address: publicKey, privateKey: privateKey, network: network, subnetwork: subnetwork, data: nil)
```
Comment (luyzdeleon) [inline pocket-aion/pocket-aion/PocketAion.swift:92]: Unnecesary `try`

File: pocket-aion/pocket-aion/PocketAion.swift [hunk 4]
```diff
@@ -0,0 +1,239 @@
...
+        guard let _ = jsContext?.evaluateScript("var account = aionInstance.eth.accounts.privateKeyToAccount('\(privateKey)')") else {
+            throw PocketPluginError.transactionCreationError("Failed to create account js object")
+        }
+        
+        guard let account = jsContext?.objectForKeyedSubscript("account")?.toObject() as? [AnyHashable: Any] else {
+            throw PocketPluginError.transactionCreationError("Failed to create account object")
+        }
+        
+        if account["address"] as? String != publicKey {
+            throw PocketPluginError.transactionCreationError("Failed to create account object")
```
Comment (luyzdeleon) [inline pocket-aion/pocket-aion/PocketAion.swift]: Better error message? `Invalid address provided.`

File: pocket-aion/pocket-aion/PocketAion.swift [hunk 5]
```diff
@@ -0,0 +1,239 @@
...
+    
+    public static func createTransaction(wallet: Wallet, params: [AnyHashable : Any]) throws -> Transaction {
+        // TODO: Find a better way to do this
+        initJS()
+        
+        // Pocket Transaction
+        let pocketTx = Transaction(obj: ["network": wallet.network, "subnetwork": wallet.subnetwork])
+        
+        // JS Account
+        guard let _ = jsContext?.evaluateScript("var account = aionInstance.eth.accounts.privateKeyToAccount('\(wallet.privateKey)')") else {
```
Comment (luyzdeleon) [inline pocket-aion/pocket-aion/PocketAion.swift]: Remove unused.

File: pocket-aion/pocket-aion/PocketAion.swift [hunk 6]
```diff
@@ -0,0 +1,239 @@
...
+            throw PocketPluginError.transactionCreationError("Failed to retrieve gas price")
+        }
+        
+        guard let gas =  params["gas"] as? String else {
+            throw PocketPluginError.transactionCreationError("Failed to retrieve gas value")
+        }
+        
+        // Exception handler
+        jsContext?.exceptionHandler = { context, error in
+            print("JsContext failed with error: \(error?.toString() ?? "none")")
```
Comment (luyzdeleon) [inline pocket-aion/pocket-aion/PocketAion.swift]: Throw new  error: `Unknown error happened`.

File: pocket-aion/pocket-aion/PocketAion.swift [hunk 7]
```diff
@@ -0,0 +1,239 @@
...
+        }
+        
+        // Promise Handler
+        let promiseBlock: @convention(block) (JavaScriptCore.JSValue, JavaScriptCore.JSValue) -> () = { (error, result) in
+            // Check for errors
+            if !error.isNull {
+                print("Failed to sign transaction with error: \(error)")
```
Comment (luyzdeleon) [inline pocket-aion/pocket-aion/PocketAion.swift]: Throw error.

File: pocket-aion/pocket-aion/PocketAion.swift [hunk 8]
```diff
@@ -0,0 +1,239 @@
...
+            }else{
+                // Retrieve result object and raw transaction
+                let resultObject = result.toObject() as! [AnyHashable: Any]
+                
+                guard let rawTx = resultObject["rawTransaction"] as? String else {
+                    print("Failed to retrieve raw signed transaction")
```
Comment (luyzdeleon) [inline pocket-aion/pocket-aion/PocketAion.swift]: throw error.

File: pocket-aion/pocket-aion/PocketAion.swift [hunk 9]
```diff
@@ -0,0 +1,239 @@
...
+        var decoderParams = [AnyHashable: Any]()
+        if decoder != nil {
+            if let returnTypes = decoder!["returnTypes"] as? [String] {
+                decoderParams["return_types"] = returnTypes
+            }
+        }
+        pocketQuery.decoder = try JSON.valueToJsonPrimitive(anyValue: decoderParams)
+        
+        // Assign network
+        pocketQuery.network = network
```
Comment (luyzdeleon) [inline pocket-aion/pocket-aion/PocketAion.swift]: Remove unnecesary.

File: pocket-aion/pocket-aion/PocketAion.swift [hunk 10]
```diff
@@ -0,0 +1,239 @@
...
+        pocketQuery.subnetwork = subnetwork
```
Comment (luyzdeleon) [inline pocket-aion/pocket-aion/PocketAion.swift]: Remove unnecesary

**Review by luyzdeleon — CHANGES_REQUESTED** [review]

File: pocket-aion/pocket-aion/PocketAion.swift (hunk 2 above)
Comment (pabelnl) [reply pocket-aion/pocket-aion/PocketAion.swift:68]: Wallet initializer throws, removing the try fails

File: pocket-aion/pocket-aion/PocketAion.swift (hunk 3 above)
Comment (pabelnl) [reply pocket-aion/pocket-aion/PocketAion.swift:92]: Wallet init throws

**Review by luyzdeleon — APPROVED** [review]

**Pull request merged**